sonar-db/src/main/java/org/sonar/db/permission/PermissionRepository.java

```diff
@@ -137,23 +142,30 @@ private void updateProjectAuthorizationDate(DbSession session, @Nullable Long pr
   }
 
   public void applyPermissionTemplate(DbSession session, String templateUuid, long resourceId) {
+    applyPermissionTemplate(session, templateUuid, resourceId, null);
+  }
+
+  private void applyPermissionTemplate(DbSession session, String templateUuid, long componentId, @Nullable Long currentUserId) {
     PermissionTemplateDto permissionTemplate = dbClient.permissionTemplateDao().selectPermissionTemplateWithPermissions(session, templateUuid);
-    updateProjectAuthorizationDate(session, resourceId);
-    dbClient.roleDao().removeAllPermissions(session, resourceId);
-    List<PermissionTemplateUserDto> usersPermissions = permissionTemplate.getUsersPermissions();
-    //TODO should return an empty list if there's no user permissions
-    if (usersPermissions != null) {
-      for (PermissionTemplateUserDto userPermission : usersPermissions) {
-        insertUserPermission(resourceId, userPermission.getUserId(), userPermission.getPermission(), false, session);
-      }
-    }
-    List<PermissionTemplateGroupDto> groupsPermissions = permissionTemplate.getGroupsPermissions();
-    //TODO should return an empty list if there's no group permission
-    if (groupsPermissions != null) {
-      for (PermissionTemplateGroupDto groupPermission : groupsPermissions) {
-        Long groupId = groupPermission.getGroupId() == null ? null : groupPermission.getGroupId();
-        insertGroupPermission(resourceId, groupId, groupPermission.getPermission(), false, session);
-      }
+    updateProjectAuthorizationDate(session, componentId);
+    dbClient.roleDao().removeAllPermissions(session, componentId);
+
+    List<PermissionTemplateUserDto> usersPermissions = requireNonNull(permissionTemplate.getUsersPermissions());
```

Comment:
As discussed, selectPermissionTemplateWithPermissions should not return PermissionTemplateDto, but another Dto that contains usersPermissions, groupsPermissions and characteristics. These 3 fields must also be removed from PermissionTemplateDto.